File: Java/[2 sem] 2DWorldSimulator/code/Antelope.java
import java.util.Random;

public class Antelope extends Animal {
    private static final int ANTELOPE_STRENGTH = 4;
    private static final int ANTELOPE_INITIATIVE = 4;
    private static final char ANTELOPE_SYMBOL = 'A';
    private static final int CHANCE_OF_ESCAPE = 50;

    public Antelope(int x, int y, World world) {
        super(x, y, world, ANTELOPE_STRENGTH, ANTELOPE_INITIATIVE, ANTELOPE_SYMBOL, 0);
    }

    @Override
    public Organism collision(Organism attacker) {
        int chance = new Random().nextInt(101);
        if (chance <= CHANCE_OF_ESCAPE) {
            int[] moveX = new int[1];
            int[] moveY = new int[1];
            this.world.findFreeNeighbourCell(this.getX(), this.getY(), moveX, moveY);
            if (moveX[0] != -1 && moveY[0] != -1) {
                this.world.moveOrganism(this, moveX[0], moveY[0]);
            }
            // return attacker but does not remove antelope
            return null;
        } else {
            return super.collision(attacker);
        }
    }

    @Override
    public boolean isAntelope() {
        return true;
    }

    @Override
    public String getSpecieName() {
        return "Antelope";
    }
}

File: C(++)/[2 sem] 2DWorldSimulator/code/src/worldObjects/animals/Antelope.cpp
//
// Created by szymo on 4/18/2023.
//

#include ".headers/Antelope.h"
#include "../../utils/helpFuncs.h"
#include "iostream"

using namespace std;

#define ANTELOPE_STRENGTH 4
#define ANTELOPE_INITIATIVE 4
#define ANTELOPE_SYMBOL 'A'
#define CHANCE_OF_ESCAPE 50

Antelope::Antelope(int x, int y, World *world) : Animal(x, y, world, ANTELOPE_STRENGTH, ANTELOPE_INITIATIVE, ANTELOPE_SYMBOL) {}

Antelope::~Antelope() = default;

Organism* Antelope::collision(Organism *attacker) {
    cout << "Antelope collision!!";
    int chance = random(0, 100);
    if (chance <= CHANCE_OF_ESCAPE)
    {
        int moveX;
        int moveY;
        this->world->findFreeNeighbourCell(this->getX(), this->getY(), moveX, moveY);
        if (moveX != -1 && moveY != -1) this->world->moveOrganism(this, moveX, moveY);
        // return attacker but does not remove antelope
        return nullptr;
    }
    else {
        return Animal::collision(attacker);
    }
}

bool Antelope::isAntelope() {
    return true;
}

string Antelope::getSpecieName() const {
    return "Antelope";
}